### ISSUE TITLE
CustomChecks should not tear down the endpoint

### ISSUE THREAD

**Issue by timbussmann** [Particular/ServiceControl.Plugin.Nsb6.CustomChecks#23]
relates to https://github.com/Particular/ServiceControl.Plugin.Nsb6.CustomChecks/issues/21

When the SC queue can't be found, the plugin raises a critical exception which will tear down the endpoint.

The heartbeats plugin was changed to log warnings instead. I think we should do the same with the CustomChecks plugin.

**Comment by johnsimons** [comment]
@timbussmann the reason CustomChecks and SagaAudit prevent the endpoint from starting and blows up is because if a user forgets to configure the ServiceControl queue we consider that a mistake and want to warn the user as quickly as possible of that issue, and logging a text line in a log file is not visible enough hence easy to miss.
On the other hand heartbeats is different because we do not consider it as critical as the other two. If the heartbeats are misconfigured, ServicePulse will warn the users and they should be able to identify the problem and rectify it.
Does this make sense ?

**Comment by timbussmann** [comment]
I disagree. I don't see any of those plugins "mission critical" to any endpoint, therefore they should not be able to shut down an endpoint. If you use these plugins, you should notice the erroneous configuration fairly quickly. For the heartbeats it has shown, that logging a warning is indeed visible enough.

> On the other hand heartbeats is different because we do not consider it as critical as the other two.

how so? What makes custom checks more critical than heartbeats? What makes a plugin which is intended for debugging more critical than heartbeats?

**Comment by johnsimons** [comment]
@timbussmann I will discuss your concern with the rest of the maintainers

**Comment by WilliamBZA** [comment]
By making misconfigured endpoints blow up, it makes it harder for ops guys to misconfigure services and not notice that anything is wrong.

On the other hand there might be custom checks that are truly inconsequential and users wouldn't mind if warnings were simply logged.

I think perhaps that there are 2 classes of CustomChecks at play here then. Perhaps we can make the error nature configurable? Maybe offer a `CustomCheck` as well as a `NonCriticalCustomCheck` which behave differently on errors.

**Comment by boblangley** [comment]
What if we just exited feature OnStart and skip validating the servicecontrol address is set if no custom checks are found? #25 

This allows us to not stop the endpoint if the plugin is just present, as was the case when reported by the user, but we will still stop the endpoint if custom checks are found and the servicecontrol address is not set.

**Comment by bording** [comment]
One thing that @timbussmann and I noticed when looking at this code, because of this line: https://github.com/Particular/ServiceControl.Plugin.Nsb6.CustomChecks/blob/ca75c4e4f817737195b27e05f74ffbd09a6795fb/src/ServiceControl.Plugin.CustomChecks/CustomChecks.cs#L48

That line can cause a critical error to be raised from a FeatureStartupTask, but instead it should just throw if it wants to stop the endpoint from starting. By using a critical error instead, the startup process isn't immediately halted, and the receivers are started while the critical error is in the process of shutting things down.

**Comment by johnsimons** [comment]
We have addressed calling the critical error at startup - https://github.com/Particular/ServiceControl.Plugin.Nsb6.CustomChecks/issues/25

**Comment by mikeminutillo** [comment]
I have a customer where a temporary outage tore the endpoint down because the CustomChecks plugin could not send results to the broker. The transport in this case (RabbitMQ) already has it's own circuit breaker for dealing with connectivity issues. In this case, should the Custom Checks simply log an error and fail quietly? Or should it tear down the endpoint?

NOTE: To get to this point we have already established that everything is configured properly and that we _can_ send messages to ServiceControl. 

@Particular/servicecontrol-maintainers

**Comment by boblangley** [comment]
Perhaps we should make the users explicitly choose the behavior? That way it isn't a surprise if a check is missing or it stops the endpoint.
